ambry-protocol/src/test/java/com.github.ambry.protocol/RequestResponseTest.java

```diff
@@ -88,66 +87,48 @@ public long getBytesRead() {
-  }
-
-  protected int sizeExcludingBlobSize() {
-    // size of (header + blobId + blob properties + metadata size + metadata + blob size + blob type)
-    return super.sizeExcludingBlobSize() - Blob_Size_InBytes - BlobType_Size_InBytes;
-  }
+class InvalidVersionPutRequest extends PutRequest {
+  static final short Put_Request_Invalid_version = 0;
 
-  @Override
-  public long writeTo(WritableByteChannel channel)
-      throws IOException {
-    long totalWritten = 0;
-    if (bufferToSend == null) {
-      bufferToSend = ByteBuffer.allocate(sizeExcludingBlobSize());
-      writeHeader();
-      bufferToSend.put(blobId.toBytes());
-      BlobPropertiesSerDe.putBlobPropertiesToBuffer(bufferToSend, properties);
-      bufferToSend.putInt(usermetadata.capacity());
-      bufferToSend.put(usermetadata);
-      bufferToSend.flip();
-    }
-    while (sentBytes < sizeInBytes()) {
-      if (bufferToSend.remaining() > 0) {
-        int toWrite = bufferToSend.remaining();
-        int written = channel.write(bufferToSend);
-        totalWritten += written;
-        sentBytes += written;
-        if (toWrite != written || sentBytes == sizeInBytes()) {
-          break;
-        }
-      }
-      logger.trace("sent Bytes from Put Request {}", sentBytes);
-      bufferToSend.clear();
-      int streamReadCount = blobStream
-          .read(bufferToSend.array(), 0, (int) Math.min(bufferToSend.capacity(), (sizeInBytes() - sentBytes)));
-      bufferToSend.limit(streamReadCount);
-    }
-    return totalWritten;
+  public InvalidVersionPutRequest(int correlationId, String clientId, BlobId blobId, BlobProperties properties,
+      ByteBuffer usermetadata, ByteBuffer blob, long blobSize, BlobType blobType) {
+    super(correlationId, clientId, blobId, properties, usermetadata, blob, blobSize, blobType);
+    versionId = Put_Request_Invalid_version;
   }
 }
 
 public class RequestResponseTest {
+  private final Random random = new Random();
+  private int testCount = 0;
+
   private void testPutRequest(MockClusterMap clusterMap, int correlationId, String clientId, BlobId blobId,
       BlobProperties blobProperties, byte[] userMetadata, BlobType blobType, byte[] blob, int blobSize)
       throws IOException {
+    testCount++;
     int sizeInBlobProperties = (int) blobProperties.getBlobSize();
     PutRequest request = new PutRequest(correlationId, clientId, blobId, blobProperties, ByteBuffer.wrap(userMetadata),
-        new ByteBufferInputStream(ByteBuffer.wrap(blob)), blobSize, blobType);
+        ByteBuffer.wrap(blob), blobSize, blobType);
     ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
-    WritableByteChannel writableByteChannel = Channels.newChannel(outputStream);
+    // In every alternate iteration, create channels in such a way that writeTo() may not be able to write out all the
+    // data at once.
+    int allocationSize =
```

Comment:
still a little unsure why we can't abstract this and run it two times - once each with `allocationSize` <> request size